appmap/_implementation/configuration.py

```diff
@@ -55,147 +56,137 @@ def packages(self):
-                  ' which %s'
-                  ' ret %s'),
-                 class_name,
-                 fn_name,
-                 which,
-                 ret)
-    return ret
+def splitname(obj):
+    """
+    Given an object that has a __module__ and __qualname__,
+    return a list of name components from both.
+    """
+    return obj.__module__.split('.') + obj.__qualname__.split('.')
 
 
-class ConfigFilter(Filter):
+def startswith(needle, haystack):
     """
-    During initialization, the ConfigFilter class reads the `package`
-    entries from the config file.  Based on these entries, it creates
-    a set of paths that should be included, and another set of paths
-    that should be excluded.
-
-    When checking to see if a function should be included,
-    ConfigFilter looks for the function in both sets.  If the function
-    is found in the included set, it's wrapped for instrumentation.
-    If it's found in the excluded set, it's used without modification.
-    If it isn't found in either set, ConfigFilter passes the function
-    on to the next Filter.
-
-    To determine whether a function is in one of the sets, the
-    fully-qualified name of the function is considered, including the
-    packages(s), module, and class that contain it.  As an example,
-    consider a function named `pkg1.pkg2.mod1.Class1.func`.  Any of
-    `pkg1`, `pkg1.pkg2.`, `pkg1.pkg2.mod1`, `pkg1.pkg2.mod1.Class1`,
-    or `pkg1.pkg2.mod1.Class1.func` will match this function (causing
-    it to be included or excluded, as appropriate).  Note that only
-    full names will match.  (This is achieved in the code by appending
-    `.` to the name retrieved from the config file.)
+    Check if needle is a prefix of haystack.
     """
-    def __init__(self, *args, **kwargs):
+    return len(needle) <= len(haystack) and all(a == b for a, b in zip(haystack, needle))
+
+
+class PathMatcher:
+    def __init__(self, prefix, excludes=None, shallow=False):
+        excludes = excludes or []
+        self.prefix = []
+        if prefix:
+            self.prefix = prefix.split('.')
+        self.excludes = [x.split('.') for x in excludes]
+        self.shallow = shallow
+
+    def matches(self, obj):
+        name = splitname(obj)
+        if startswith(self.prefix, name):
+            name = name[len(self.prefix):]
+            result = not any(startswith(x, name) for x in self.excludes)
+        else:
+            result = False
+        logger.debug('%r.matches(%r) -> %r', self, obj, result)
+        return result
+
+    def __repr__(self):
+        return 'PathMatcher(%r, %r, shallow=%r)' % (
+            '.'.join(self.prefix),
+            ['.'.join(ex) for ex in self.excludes],
+            self.shallow
+        )
+
+
+class DistMatcher(PathMatcher):
+    def __init__(self, dist, *args, **kwargs):
         super().__init__(*args, **kwargs)
+        self.dist = dist
+        self.files = [str(pp.locate()) for pp in importlib_metadata.files(dist)]
```

Comment:
Why do the matching on file rather than on module?

Reply:
Because packages don't provide modules, packages provide files. One package can provide many modules and the same module can be provided by many packages (in which case of course only one implementation can be loaded at the same time, at least at the root namespace). A module doesn't know its package; in fact, it doesn't know about packages at all (and neither does python core). There's no way to map from module to package except by matching its file path to the package manifest.

Reply:
Ok, I understand this now. Thanks.